### PR TITLE
fix <amp-ad> <amp-ad-3p-impl> collapse issue

### PR DESCRIPTION
**Pull request by zhouyx**
I tried to layout=fill for <amp-ad-3p-impl> and not copy height/width/heights/layout attributes from <amp-ad>. So later <amp-ad> could take care of resizing for its child <amp-ad-3p-impl> 

changes: 
copy <amp-ad> children element to <amp-ad-3p-impl>
expand BOOKKEEPING_ATTRIBUTES values to prevent layout attribute copy
change <amp-ad-3p-impl> layout type to fill
override AmpAd3PImpl `attemptChangeHeight` and `attemptChangeSize` function

### CODE REVIEW

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 1]
```diff
@@ -156,7 +157,9 @@ export class AmpAd3PImpl extends AMP.BaseElement {
 
   /** @override */
   isLayoutSupported(layout) {
-    return isLayoutSizeDefined(layout);
+    // add for testing
+    //return isLayoutSizeDefined(layout);
+    return layout == Layout.FILL;
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: check for layout=fill

File: extensions/amp-ad/0.1/amp-ad.js [hunk 2]
```diff
@@ -52,6 +52,18 @@ function copyAttributes(sourceElement, targetElement) {
     }
   }
 }
+//TODO(zhouyx) add for first version testing
+function copyAttributesLayoutFill(sourceElement, targetElement) {
+  const attrs = sourceElement.attributes;
+  for (let i = attrs.length - 1; i >= 0; --i) {
+    const attr = attrs[i];
+    if (!BOOKKEEPING_ATTRIBUTES_.hasOwnProperty(attr.name)) {
+      if (attr.name != 'height' && attr.name != 'width' && attr.name != 'layout' && attr.name != 'heights') {
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad.js]: not copying `height/width/layout/heights` attributes

File: src/3p-frame.js [hunk 3]
```diff
@@ -117,8 +117,18 @@ export function getIframe(parentWindow, element, opt_type) {
...
+  // console.log('height', attributes.height);
+  if (attributes.hasOwnProperty('width')) {
+    // console.log('in')
+    iframe.width = attributes.width;
+  }
+  // TODO: <amp-ad-3p-impl> doesn't get width and height, what to do here?
+
+  if (attributes.hasOwnProperty('height')) {
+    iframe.height = attributes.height;
+  }
```
Comment (zhouyx) [inline src/3p-frame.js]: I am thinking of getting width and height from element.parent(which is amp-ad) directly here?

**Comment by zhouyx** [comment]
By setting `<amp-ad-3p-impl>` layout to fill, and not copying `width` `height` from `<amp-ad>`, the `iframe` attributes are not set correctly. I think we need to figure a way to skip `<amp-ad-3p-impl>` and copy `<amp-ad>` attribute directly to iframe.

**Comment by zhouyx** [comment]
@dvoytenko This is the `layout=fill` change

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 4]
```diff
@@ -352,4 +357,12 @@ export class AmpAd3PImpl extends AMP.BaseElement {
         overflown, requestedHeight, requestedWidth);
     }
   }
+
+  /** Function to collapse parent <amp-ad> element*/
+  collapseParent() {
+    this.resources_.attemptChangeSize(this.parent_, 0, 0, () => {
+      this.parent_.style.display = 'none';
```
Comment (cramforce) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: This should be enough :)

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 5]
```diff
@@ -352,4 +357,12 @@ export class AmpAd3PImpl extends AMP.BaseElement {
         overflown, requestedHeight, requestedWidth);
     }
   }
+
+  /** Function to collapse parent <amp-ad> element*/
+  collapseParent() {
+    this.resources_.attemptChangeSize(this.parent_, 0, 0, () => {
```
Comment (cramforce) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: You also need to fix
https://github.com/ampproject/amphtml/blob/master/extensions/amp-ad/0.1/amp-ad-api-handler.js#L147

Maybe you can actually fix both by overriding `attemptChangeHeight` in this class to operate on the parent.

File: extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js [hunk 6]
```diff
@@ -17,7 +17,7 @@
 import {createAdPromise} from '../../../../testing/ad-iframe';
 import * as sinon from 'sinon';
 
-describe('amp-ad-3p-impl', tests('amp-ad-3p-impl'));
+describe.only('amp-ad-3p-impl', tests('amp-ad-3p-impl'));
```
Comment (zhouyx) [inline extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js]: 😢 Forget to delete only ; (

File: extensions/amp-ad/0.1/amp-ad.js [hunk 7]
```diff
@@ -88,6 +99,8 @@ export class AmpAd extends AMP.BaseElement {
       /*OK*/insertAmpExtensionScript(this.getWin(), extensionTag, true);
     }
     copyAttributes(this.element, newChild);
+    newChild.setAttribute('layout', 'fill');
```
Comment (cramforce) [inline extensions/amp-ad/0.1/amp-ad.js]: I don't follow the question.

**Comment by cramforce** [comment]
This looks great!

File: extensions/amp-ad/0.1/amp-ad.js (hunk 7 above)
Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad.js]: `const extensionTag = networkImplementationTag(type);` Don't quite understand this line of code. And I'm not sure if we will also create a `AmpAd3PImpl` element

Comment (cramforce) [inline extensions/amp-ad/0.1/amp-ad.js]: Ah, no need to worry about that right now. The A4A team can :)

**Comment by zhouyx** [comment]
I am looking at #3962, with that do we still need the fix?

**Comment by cramforce** [comment]
It is good to have it in our back pocket :)

But it would be better to land #3962

File: extensions/amp-ad/0.1/amp-ad.js (hunk 7 above)
Comment (tdrl) [inline extensions/amp-ad/0.1/amp-ad.js]: That constructs the name of the extension tag.  So, networkImplementationTag('adsense') => 'amp-ad-network-adsense-impl'.  It's used both as the tag name and extension name for requesting dymamic extension load.

File: extensions/amp-ad/0.1/amp-ad.js [hunk 8]
```diff
@@ -33,7 +33,8 @@ function networkImplementationTag(type) {
 }
 
 /** @private @enum {!number} */
-const BOOKKEEPING_ATTRIBUTES_ = {'class': 1, 'style': 2, 'id': 3};
+const BOOKKEEPING_ATTRIBUTES_ = {'class': 1, 'style': 2, 'id': 3, 'layout': 4,
+    'width': 5, 'height': 6, 'heights': 7, 'sizes': 8, 'media': 9};
```
Comment (tdrl) [inline extensions/amp-ad/0.1/amp-ad.js]: Shouldn't width, height, etc. get passed on to the child tag?  How else will they know what the requested sizes/media are?

Comment (zhouyx) [inline extensions/amp-ad/0.1/amp-ad.js]: My approach here is let the parent `<amp-ad>` decide the sizes and layout

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 9]
```diff
@@ -352,4 +359,16 @@ export class AmpAd3PImpl extends AMP.BaseElement {
         overflown, requestedHeight, requestedWidth);
     }
   }
+
+  /** @override */
+  attemptChangeHeight(newHeight, opt_callback) {
+    this.resources_.attemptChangeSize(this.parent_, newHeight,
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Can't you `this.parent_.attempChangeHeight(newHeight, opt_callback)`?

File: extensions/amp-ad/0.1/amp-ad.js [hunk 10]
```diff
@@ -50,6 +51,16 @@ function copyAttributes(sourceElement, targetElement) {
...
       targetElement.setAttribute(attr.name, attr.value);
     }
+    // special case put here for width and height
+    if (attr.name == 'width' || attr.name == 'height') {
+      targetElement.setAttribute(attr.name, attr.value);
+    }
+  }
+}
+
+function moveChildren(sourceElement, targetElement) {
```
Comment (jridgewell) [inline extensions/amp-ad/0.1/amp-ad.js]: This should be in the `dom` module.

**Comment by cramforce** [comment]
CL is no longer needed due to #3962

**Pull request closed without merging**